service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java

```diff
@@ -312,6 +312,36 @@ public void deleteGcpCloudContext(UUID workspaceId, AuthenticatedUserRequest use
         .submitAndWait(null);
   }
 
+  /**
+   * Retrieve the optional GCP cloud context
+   *
+   * @param workspaceId workspace identifier of the cloud context
+   * @return optional GCP cloud context
+   */
+  public Optional<GcpCloudContext> getGcpCloudContext(UUID workspaceId) {
+    return workspaceDao.getGcpCloudContext(workspaceId);
+  }
+
+  /**
+   * Retrieve the optional GCP cloud context, providing a workspace This is a frequent usage, so we
+   * make a method for it to save coding the fetch of workspace id every time
+   */
+  public Optional<GcpCloudContext> getGcpCloudContext(Workspace workspace) {
+    return workspaceDao.getGcpCloudContext(workspace.getWorkspaceId());
+  }
+
+  /**
+   * Helper method for looking up the GCP project ID for a given workspace ID, if one exists. Unlike
+   * {@link #getRequiredGcpProject(UUID)}, this returns an empty Optional instead of throwing if the
+   * given workspace does not have a GCP cloud context.
+   *
+   * @param workspaceId workspace identifier of the cloud context
+   * @return optional GCP project from the cloud context
+   */
+  public Optional<String> getGcpProject(UUID workspaceId) {
```

Comment:
So a workspace will have 0 or 1 GCP projects and 0 or 1 Azure projects?

Reply:
Right